### PR TITLE
#276 - added SECP256K1 hierarchical key derivation utilities

### PR DESCRIPTION
**Pull request by thehenrytsai**
* added SECP256K1 hierarchical key derivation utilities
* refactored existing SECP256K1 code for better reusability

Note that only 4 public methods are added in this PR (all starts with `derive`), the rest of the methods/code were already there, just refactored.

### CODE REVIEW

File: src/utils/secp256k1.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+  /**
+   * Parses the given key derivation path.
+   * @returns Path segments if successfully validate the derivation path.
+   * @throws {DwnError} with `DwnErrorCode.HdKeyDerivationPathInvalid` if derivation path fails validation.
+   */
+  private static parseAndValidateKeyDerivationPath(derivationPath: string): string[] {
+    const pathSegments = derivationPath.split('/');
+
+    if (pathSegments.length === 0 || pathSegments.includes('')) {
```
Comment (diehuxx) [inline src/utils/secp256k1.ts:201]: Do we need to validate that path segments are in ASCII or anything like that?

File: tests/jose/algorithms/signing/secp256k1.spec.ts
Comment (diehuxx) [inline tests/jose/algorithms/signing/secp256k1.spec.ts:1]: We should move this test file `tests/utils/secp256k1.spec.ts`

File: tests/jose/algorithms/signing/secp256k1.spec.ts [hunk 2]
```diff
@@ -1,31 +1,93 @@
...
+      const { publicKey, privateKey } = await Secp256k1.generateKeyPairRaw();
+
+      const path = '1/2/3/4/5/6/7/8/9/10';
+
+      const derivedKeyFromPublicKey = await Secp256k1.derivePublicKey(publicKey, path);
+      const derivedKeyFromPrivateKey = await Secp256k1.derivePublicKey(privateKey, path);
+      expect(Comparer.byteArraysEqual(derivedKeyFromPublicKey, derivedKeyFromPrivateKey)).to.be.true;
+    });
+
+    it('should derive the same public key using either the private or public counterpart of the same key pair', async () => {
```
Comment (diehuxx) [inline tests/jose/algorithms/signing/secp256k1.spec.ts:84]: Related to my other comment, should we also test derivation path validation for paths that include like emojis and empty strings?

File: src/utils/secp256k1.ts (hunk 1 above)
Comment (thehenrytsai) [reply src/utils/secp256k1.ts:201]: No, in fact, I have a subsequent PR that changed path to be an array of strings so that the path has less limitation (such as delimiter by `/`). I'll see if I can easily move that change into this PR. If not, you'll see it in the upcoming one.

File: tests/jose/algorithms/signing/secp256k1.spec.ts (hunk 2 above)
Comment (diehuxx) [reply tests/jose/algorithms/signing/secp256k1.spec.ts:84]: resolving based on your reply above. ty

File: tests/jose/algorithms/signing/secp256k1.spec.ts
Comment (thehenrytsai) [reply tests/jose/algorithms/signing/secp256k1.spec.ts:1]: Good suggestion!

**Review by diehuxx — APPROVED** [review]
